File: model/src/main/java/io/github/chriszhong/model/organization/event/PolicyEventFactory.java
package io.github.chriszhong.model.organization.event;

import io.github.chriszhong.model.organization.entity.Policy;

/**
 * The {@linkplain PolicyEventFactory} interface defines the API for constructing {@linkplain PolicyEvent}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface PolicyEventFactory {

	/**
	 * Constructs an {@linkplain PolicyEvent}.
	 *
	 * @param category
	 *            the {@linkplain EventCategory}.
	 * @param policy
	 *            the {@linkplain Policy}.
	 *
	 * @return a {@linkplain PolicyEvent}.
	 */
	PolicyEvent build(EventCategory category, Policy policy);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/PolicyEventFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Policy
import runtimemodels.chazm.model.event.EventType

/**
 * The [PolicyEventFactory] interface defines the API for constructing [PolicyEvent]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface PolicyEventFactory {
    /**
     * Constructs an [PolicyEvent].
     *
     * @param category the [EventType].
     * @param policy   the [Policy].
     * @return a [PolicyEvent].
     */
    fun build(category: EventType, policy: Policy): PolicyEvent
}
